File: backend/worktracker/models/launch_binding.py
from django.core.exceptions import ValidationError
from django.db import models

from .issue_type import IssueType
from .provider_catalog import AgentModel, ReasoningLevel
from .state import State


class LaunchBinding(models.Model):
    """Explicit agent launch policy for one work-item type/current-state pair."""

    issue_type = models.ForeignKey(
        IssueType, on_delete=models.CASCADE, related_name="launch_bindings"
    )
    state = models.ForeignKey(
        State, on_delete=models.CASCADE, related_name="launch_bindings"
    )
    prompt = models.TextField(blank=True, default="")
    required_skills = models.JSONField(default=list, blank=True)
    model = models.ForeignKey(
        AgentModel,
        on_delete=models.PROTECT,
        related_name="launch_bindings",
        blank=True,
        null=True,
    )
    reasoning = models.ForeignKey(
        ReasoningLevel,
        on_delete=models.PROTECT,
        related_name="launch_bindings",
        blank=True,
        null=True,
    )
    auto_start = models.BooleanField(default=False)
    subtree_run_enabled = models.BooleanField(default=False)
    created_at = models.DateTimeField(auto_now_add=True)
    updated_at = models.DateTimeField(auto_now=True)

    class Meta:
        constraints = [
            models.UniqueConstraint(
                fields=("issue_type", "state"),
                name="unique_launch_binding_type_state",
            )
        ]
        ordering = ("issue_type__sort_order", "state__sort_order", "id")

    @property
    def has_launch_policy(self) -> bool:
        """Whether this row carries launch policy rather than only a flag.

        Automation flags no longer create flag-only rows. Every "is a launch
        configured here" read still goes through this predicate because the
        prompt can deliberately inherit its model from the global default.
        """

        return bool(self.prompt.strip() or self.model_id)

    @property
    def provider_slug(self) -> str | None:
        """Return the provider implied by the selected catalog model."""

        return self.model.provider.slug if self.model_id else None

    @property
    def model_name(self) -> str | None:
        return self.model.name if self.model_id else None

    @property
    def reasoning_name(self) -> str | None:
        return self.reasoning.name if self.reasoning_id else None

    def clean(self):
        super().clean()
        errors = {}
        if (
            self.issue_type_id
            and self.state_id
            and self.issue_type.project_id != self.state.project_id
        ):
            errors["state"] = (
                "Current state and work-item type must belong to the same project."
            )

        from worktracker.services.launch_bindings import (  # avoid import cycle
            LaunchBindingError,
            validate_required_skills,
        )

        if self.reasoning_id and not self.model_id:
            errors["reasoning"] = "Choose a model before configuring reasoning."
        elif (
            self.reasoning_id
            and self.model_id
            and not self.model.permitted_reasoning_levels.filter(
                pk=self.reasoning_id
            ).exists()
        ):
            errors["reasoning"] = (
                f"Reasoning '{self.reasoning.name}' is not permitted for model "
                f"'{self.model.name}'."
            )
        if (self.auto_start or self.subtree_run_enabled) and not self.has_launch_policy:
            errors["auto_start"] = (
                "Configure a launch binding before enabling automation."
            )
        try:
            self.required_skills = validate_required_skills(
                required_skills=self.required_skills,
                prompt=self.prompt,
            )
        except LaunchBindingError as exc:
            errors[exc.field or "required_skills"] = exc.message
        if errors:
            raise ValidationError(errors)

File: studio/src-tauri/src/entities/work_management/launch_binding.rs
use sea_orm::entity::prelude::*;

#[sea_orm::model]
#[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
#[sea_orm(table_name = "worktracker_launchbinding")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub id: i64,
    pub issue_type_id: String,
    pub state_id: String,
    pub prompt: String,
    pub required_skills: Json,
    /// The one skill a launch enters through, or none.
    ///
    /// Nullable, and when present it names a skill the binding already
    /// requires. The stored value is the bare skill slug: provider prefix
    /// characters belong to the command a launch builds, not to this row.
    pub entry_skill: Option<String>,
    pub model_id: Option<String>,
    pub reasoning_id: Option<String>,
    pub auto_start: bool,
    pub subtree_run_enabled: bool,
    pub created_at: DateTime,
    pub updated_at: DateTime,
    #[sea_orm(belongs_to, from = "issue_type_id", to = "id")]
    pub issue_type: BelongsTo<super::issue_type::Entity>,
    #[sea_orm(belongs_to, from = "state_id", to = "id")]
    pub state: BelongsTo<super::state::Entity>,
    #[sea_orm(belongs_to, from = "model_id", to = "id")]
    pub agent_model: BelongsTo<Option<super::agent_model::Entity>>,
    #[sea_orm(belongs_to, from = "reasoning_id", to = "id")]
    pub reasoning: BelongsTo<Option<super::reasoning_level::Entity>>,
}

impl ActiveModelBehavior for ActiveModel {}
